contracts/LRTDepositPool.sol

```diff
@@ -164,13 +177,95 @@ contract LRTDepositPool is ILRTDepositPool, LRTConfigRoleChecker, PausableUpgrad
     }
 
     /// @dev private function to mint primeETH
-    /// @param primeEthAmount Amount of primeETH minted
-    function _mint(uint256 primeEthAmount) private {
+    /// @param amount Amount of primeETH to be minted
+    function _mint(uint256 amount) private {
         address primeETH = lrtConfig.primeETH();
         // mint primeETH for user
-        IPrimeETH(primeETH).mint(msg.sender, primeEthAmount);
+        IPrimeETH(primeETH).mint(msg.sender, amount);
     }
 
+    /*//////////////////////////////////////////////////////////////
+                        Withdraw functions
+    //////////////////////////////////////////////////////////////*/
+
+    /// @notice PrimeETH staker requests the withdrawal of a liquid staking token (LST) from
+    // its respective underlying EigenLayer strategy.
+    /// @dev Will emit the `Withdrawal` event from EigenLayer's `DelegationManager` contract
+    /// which is needed for the `claimWithdrawal` call.
+    /// @param asset address of the liquid staking token (LST) being requested. eg OETH. Can not be WETH.
+    /// @param assetAmount the amount of LSTs to withdraw.
+    /// @param maxPrimeETH the maximum amount of primeETH tokens that can be burned.
+    /// @return primeETHAmount the amount of primeETH tokens that were burned.
+    function requestWithdrawal(
+        address asset,
+        uint256 assetAmount,
+        uint256 maxPrimeETH
+    )
+        external
+        whenNotPaused
+        nonReentrant
+        returns (uint256 primeETHAmount)
+    {
+        if (assetAmount == 0) {
+            revert ZeroAmount();
+        }
+
+        // Can only withdraw the supported asset
+        if (asset != WITHDRAW_ASSET) {
+            revert NotWithdrawAsset();
+        }
+
+        // Convert asset amount to primeETH amount.
+        // Using mint function here even though it's a withdrawal as it does the same calculation
+        primeETHAmount = getMintAmount(asset, assetAmount);
```

Comment:
🔴 This rounds down. The user essentially always burns less prime ETH than they withdraw in assets.

<img width="343" alt="image" src="https://github.com/oplabs/primestaked-eth/assets/837/65095b5a-b15a-45b1-9e54-73bc7f845958">

A withdraw of raw 1 will burn 0 prime ETH. I suspect that a user can withdraw funds from the system without owning any prime ETH at all. At the least we are going to hear about this from bug bounty people!

What if we flip this method over to work with a `primeEthAmount` argument instead of a `assetAmount` argument. We then just use the prime eth exchange rate to calculate the number of assets the user will receive (and do so in a way that will round down.)

We'll need a comment that the only allowed withdraw assets should be 1 to 1 with the price of ETH with no external oracles, but that currently a needed security requirement anyway.

This also handles better the case of a user wanting to withdraw all their funds. No calculation now, just withdraw the balance.

Reply:
Good pickup. I was always nervous about reusing `getMintAmount` but missed the rounding issue.

How about we round up outside the `getMintAmount` call?

```Solidity
// Convert asset amount to primeETH amount.
// Using mint function here even though it's a withdrawal as it does the same calculation.
// We round up the primeETH amount for withdrawals as the the divide by the primeETH exchange rate will truncate
// down.
primeETHAmount = getMintAmount(asset, assetAmount) + 1;
```